File: plugin/src/main/java/org/octopusden/octopus/jira/api/rest/ApiRestController.java
package org.octopusden.octopus.jira.api.rest;

import com.atlassian.plugins.rest.common.security.AnonymousAllowed;
import com.fasterxml.jackson.databind.ObjectMapper;
import org.octopusden.octopus.jira.api.dto.IPSRequest;
import org.octopusden.octopus.jira.api.exception.BadRequestException;
import org.octopusden.octopus.jira.api.exception.FailedGenerateIPSException;
import org.octopusden.octopus.jira.api.service.IPSService;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import javax.ws.rs.DefaultValue;
import javax.ws.rs.GET;
import javax.ws.rs.Path;
import javax.ws.rs.PathParam;
import javax.ws.rs.Produces;
import javax.ws.rs.QueryParam;
import javax.ws.rs.core.MediaType;
import javax.ws.rs.core.Response;
import java.time.LocalDate;
import java.time.Year;
import java.time.ZoneOffset;
import java.time.format.DateTimeFormatter;
import java.time.format.DateTimeParseException;
import java.util.Date;

@AnonymousAllowed
@Path("/api")
public class ApiRestController {

    private static final Logger logger = LoggerFactory.getLogger(ApiRestController.class);
    private static final DateTimeFormatter DATE_FORMAT_YEAR = DateTimeFormatter.ofPattern("yyyy");
    private static final DateTimeFormatter DATE_FORMAT_DATE = DateTimeFormatter.ofPattern("yyyy-MM-dd");

    private final IPSService ipsService;
    private final ObjectMapper objectMapper;

    public ApiRestController(IPSService ipsService, ObjectMapper objectMapper) {
        this.ipsService = ipsService;
        this.objectMapper = objectMapper;
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @Path("ips/{ips}")
    public Response getIPS(
            @PathParam("ips") String ips,
            @QueryParam("since") Integer sinceYear,
            @QueryParam("sinceDate") String sinceDate,
            @QueryParam("release") String release,
            @QueryParam("system") @DefaultValue("CLASSIC") String system,
            @QueryParam("mandatory") @DefaultValue("true") boolean mandatory
    ) {
        IPSRequest request;
        if (release != null) {
            request = new IPSRequest(ips, release, null, system, mandatory);
        } else {
            Date startDate;
            try {
                if (sinceYear != null) {
                    startDate = Date.from(
                            Year.parse(String.valueOf(sinceYear), DATE_FORMAT_YEAR)
                                    .atDay(1).atStartOfDay().toInstant(ZoneOffset.UTC)
                    );
                } else if (sinceDate != null) {
                    startDate = Date.from(
                            LocalDate.parse(sinceDate, DATE_FORMAT_DATE)
                                    .atStartOfDay().toInstant(ZoneOffset.UTC)
                    );
                } else {
                    throw new BadRequestException("Either since or sinceDate parameters need to be set");
                }
            } catch (DateTimeParseException e) {
                throw new BadRequestException("Invalid date format: " + e.getMessage());
            }
            request = new IPSRequest(ips, null, startDate, system, mandatory);
        }

        try {
            Object result = ipsService.generate(request);
            logger.info("Generated IPS data for {}:{}", request.getIps(), request.getRelease());
            return Response.ok(objectMapper.writeValueAsString(result)).build();
        } catch (IllegalArgumentException e) {
            String msg = "Bad request generating IPS data for " + request.getIps() + ":" + request.getRelease()
                    + " — " + (e.getMessage() != null ? e.getMessage() : "unknown error");
            logger.warn(msg);
            throw new BadRequestException(msg);
        } catch (Exception e) {
            String msg = "Failed to generate IPS data for " + request.getIps() + ":" + request.getRelease() + " " + e;
            logger.error(msg);
            throw new FailedGenerateIPSException(msg);
        }
    }
}

File: plugin/src/main/kotlin/org/octopusden/octopus/jira/api/rest/ApiRestController.kt
package org.octopusden.octopus.jira.api.rest

import com.atlassian.plugins.rest.common.security.AnonymousAllowed
import com.fasterxml.jackson.databind.ObjectMapper
import org.octopusden.octopus.jira.api.dto.IPSRequest
import org.octopusden.octopus.jira.api.exception.BadRequestException
import org.octopusden.octopus.jira.api.exception.FailedGenerateIPSException
import org.octopusden.octopus.jira.api.service.IPSService
import java.net.URLDecoder
import java.time.LocalDate
import java.time.ZoneOffset
import java.time.format.DateTimeFormatter
import java.time.format.DateTimeParseException
import java.util.Date
import javax.ws.rs.DefaultValue
import javax.ws.rs.GET
import javax.ws.rs.Path
import javax.ws.rs.PathParam
import javax.ws.rs.Produces
import javax.ws.rs.QueryParam
import javax.ws.rs.core.MediaType
import javax.ws.rs.core.Response
import org.slf4j.LoggerFactory
import java.time.Year

@AnonymousAllowed
@Path("/api")
class ApiRestController(
    private val ipsService: IPSService,
    private val objectMapper: ObjectMapper,
) {
    companion object {
        private val logger = LoggerFactory.getLogger(ApiRestController::class.java)
        private val DATE_FORMAT_YEAR = DateTimeFormatter.ofPattern("yyyy")
        private val DATE_FORMAT_DATE = DateTimeFormatter.ofPattern("yyyy-MM-dd")
    }

    @GET
    @Produces(MediaType.APPLICATION_JSON)
    @Path("ips/{ips}")
    fun getIPS(
        @PathParam("ips") ips: String,
        @QueryParam("since") sinceYear: Int?,
        @QueryParam("sinceDate") sinceDate: String?,
        @QueryParam("release") release: String?,
        @QueryParam("system") @DefaultValue("CLASSIC") system: String,
        @QueryParam("mandatory") @DefaultValue("true") mandatory: Boolean
    ): Response {
        val request = if (release != null) {
            IPSRequest(
                ips = URLDecoder.decode(ips, "UTF-8"),
                release = release,
                startDate = null,
                system = system,
                mandatory = mandatory
            )
        } else {
            val startDate = try {
                when {
                    sinceYear != null -> Date.from(
                        Year.parse(sinceYear.toString(), DATE_FORMAT_YEAR)
                            .atDay(1).atStartOfDay().toInstant(ZoneOffset.UTC)
                    )
                    sinceDate != null -> Date.from(
                        LocalDate.parse(sinceDate, DATE_FORMAT_DATE)
                            .atStartOfDay().toInstant(ZoneOffset.UTC)
                    )
                    else -> throw BadRequestException("Either since or sinceDate parameters need to be set")
                }
            } catch (e: DateTimeParseException) {
                throw BadRequestException("Invalid date format: ${e.message}")
            }
            IPSRequest(
                ips = URLDecoder.decode(ips, "UTF-8"),
                release = null,
                startDate = startDate,
                system = system,
                mandatory = mandatory
            )
        }

        return try {
            val result = ipsService.generate(request)
            logger.info("Generated IPS data for ${request.ips}:${request.release}")
            Response.ok(objectMapper.writeValueAsString(result)).build()
        } catch (e: IllegalArgumentException) {
            "Bad request generating IPS data for ${request.ips}:${request.release} — ${e.message ?: "unknown error"}".let {
                logger.warn(it)
                throw BadRequestException(it)
            }
        } catch (e: Exception) {
            "Failed to generate IPS data for ${request.ips}:${request.release} ${e}".let {
                logger.error(it)
                throw FailedGenerateIPSException(it)
            }
        }
    }
}
